lib/Dialect/HW/HWOps.cpp

```diff
@@ -1613,6 +1613,19 @@ void StructExtractOp::build(OpBuilder &builder, OperationState &odsState,
   build(builder, odsState, resultType, input, fieldAttr);
 }
 
+// A struct extract of a struct create -> corresponding struct create operand.
+OpFoldResult StructExtractOp::fold(ArrayRef<Attribute> operands) {
+  auto structCreate = dyn_cast_or_null<StructCreateOp>(input().getDefiningOp());
+  if (!structCreate)
+    return {};
+  auto ty = type_dyn_cast<StructType>(input().getType());
```

Comment:
This has to be true, yes?  If so then just cast<StructType>

Reply:
My thinking was that we probably never want a folder to assert/crash, even in the presence of invalid IR so I coded this on the conservative side.

Reply:
OTOH invalid IR would more likely have a null `input()`. And no other folder checks for that, so I'll switch.

Reply:
Structural validity is checked by the verifiers so that passes don't have to be paranoid.  As long as the td def limits this to a struct type, you shouldn't redundantly check it.